packages/plugin-hardhat/src/utils/options.ts

```diff
@@ -25,7 +29,8 @@ export function withDefaults(opts: UpgradeOptions = {}): Required<UpgradeOptions
     constructorArgs: opts.constructorArgs ?? [],
     timeout: opts.timeout ?? 60e3,
     pollingInterval: opts.pollingInterval ?? 5e3,
-    useDeployedImplementation: opts.useDeployedImplementation ?? true,
+    useDeployedImplementation: opts.useDeployedImplementation ?? false,
```

Comment:
I'm confused by a couple of things here.

1. Why would the default change?
2. Was the default really `true`? I thought our default was always the `onchange` behavior.

See my other comment. I think we should be returning an options object that is self-consistent. Can we return an object without the `useDeployedImplementation` field?

Reply:
I see now that if the user set `useDeployedImplementation` in their config we want to refer to that in the error message. This makes sense.

Perhaps the way it's implemented currently is the best option, let me know what you think. My comment was because I would expect the deprecated option to be handled in this location instead of in the other places.

Reply:
Re: original default being `true` - that was a bug but had no consequence because we never actually used the default value from `withDefaults`.

In [deployImpl](https://github.com/ericglau/openzeppelin-upgrades/blob/305828bfcfd9f73baea259ee511064a838b69ddb/packages/plugin-hardhat/src/utils/deploy-impl.ts#L116), we use the `opts` param instead of `deployData.fullOpts`, where only `deployData.fullOpts` was populated by `withDefaults`.

I agree we could remove `useDeployedImplementation` from `withDefaults`, but this function's return signature is currently `Required<UpgradeOptions>` so we would need to get around that.

Technically the default of `redeployImplementation = onchange` seems kind of pointless as well -- we could equally dictate that `undefined` means "on change" and not allow an explicit `onchange` value.  It does help to make it clearer though.  But (similar to `useDeployedImplementation` above) is there any benefit in setting `redeployImplementation = onchange` in `withDefaults` if we never use this default value?

Reply:
Leaving the defaults as-is but improved the way we consume `fullOpts` (coming from `withDefaults`) and `opts`